Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -30,167 +30,320 @@
+                    // other than 1 (currently ignored but should be added in the future).
+                    _curOperation = candidateOperation;
+                }
+
+                [candidateOperation start];
+
+                {
+                    std::lock_guard<std::mutex> lock(_lock);
+
+                    // Clear the current operation now that start is finished.
+                    // NOTE: the assumption (inccorect) here is that the operation is now finished and
+                    // we no longer need any accounting of the current operation.
+                    // This is wrong because an operation could be a new fangled asynchronous
+                    // one and it could finish muc later (not synchronously executed with [operation start]).
+                    _curOperation = nil;
+                }
+            } else {
+                {
+                    std::lock_guard<std::mutex> lock(_lock);
+                    // Not yet ready. Demote it to the holding area. Will pick it up again later.
+                    _operationsHoldingArea[priority].emplace_back(std::move(candidateOperation));
+                }
+
+                // We claim that we have done some work here. This keeps the queue pumping and
+                // will eventually bring this item back out for consideration and will continue
+                // through this list for anyother ready operations.
+                return TRUE;
+            }
+        }
 
-    NSOperationQueuePriv() {
-        memset(this, 0, sizeof(NSOperationQueuePriv));
-        _threadRunningLock = PTHREAD_MUTEX_INITIALIZER;
+        return candidateOperation != nil;
     }
-};
 
-extern pthread_key_t g_currentDispatchQueue;
+    void WaitUntilNotSuspended() {
+        std::unique_lock<std::mutex> lock(_lock);
+        while (_isSuspended) {
+            _suspendedCondition.wait(lock);
+        }
+        lock.unlock();
+    }
 
-static inline gpointer CompareExchangePointer(volatile gpointer* dest, gpointer exch, gpointer comp) {
-    return (gpointer)EbrCompareExchange((volatile int*)dest, (long)exch, (long)comp);
-}
+    void ClearCandidateLists() {
+        std::lock_guard<std::mutex> lock(_lock);
+        for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
+            _candidateOperationsToRunNext[i].clear();
+        }
+    }
 
-NSAtomicListRef NSAtomicListSteal(NSAtomicListRef* listPtr) {
-    NSAtomicListRef ret;
-    do {
-        ret = *listPtr;
-    } while (CompareExchangePointer((void**)listPtr, NULL, ret) != ret);
-    return ret;
-}
+    BOOL HasMoreWork() {
+        std::lock_guard<std::mutex> lock(_lock);
 
-void NSAtomicListReverse(NSAtomicListRef* listPtr) {
-    struct NSAtomicListNode* cur = *listPtr;
-    struct NSAtomicListNode* prev = NULL;
-    struct NSAtomicListNode* next = NULL;
+        BOOL toReturn = _HasMoreWorkNoLock();
+
+        if (!toReturn) {
```

Comment:
I'll add a note. I agree that its weird. I didn't have it like that initially but that is most in the spirit of the previous code. I very much dislike nil'ing out the thread to indicate state but some flag is needed under the lock to indicate that things are done and a new thread is needed next go around. 

A much preferable solution long term is to use a threadpool or some other mechanism where we aren't manually trying to optimize thread lifetime.

---
In reply to: [95450052](https://github.com/Microsoft/WinObjC/pull/1676#discussion_r95450052) [](ancestors = 95450052)